File: java/src/model/ataque/Ataque.java
package model.ataque;

import model.chimpos.Chinpokomon;

public abstract class Ataque {
    private Integer valor;
    private String nombre;
    private Integer valorExtra;

    public Ataque(Integer valor, String nombre, Integer valorExtra) {
        this.valor = valor;
        this.nombre = nombre;
        this.valorExtra = valorExtra;
    }

    public String getNombre() {
        return nombre;
    }

    public Integer getValor() {
        return valor;
    }

    public void setValor(Integer valor) {
        this.valor = valor;
    }

    public void setNombre(String nombre) {
        this.nombre = nombre;
    }

    public Integer getValorExtra() {
        return valorExtra;
    }

    public void setValorExtra(Integer valorExtra) {
        this.valorExtra = valorExtra;
    }

    public void atacar(Chinpokomon chipo) {
        if(chipo.getOponente().tieneVentajaSobre(chipo)) {
            this.atacar(chipo, this.getValor() + this.getValorExtra());
        } else {
            this.atacar(chipo, this.getValor());
        }
    }

    public abstract void atacar(Chinpokomon chipo, Integer valor);

}

File: python/model/ataque/Ataque.py
from abc import ABC, abstractmethod

class Ataque():
    def __init__(self, valor, nombre, valor_extra):
        self._valor = valor
        self._nombre = nombre
        self._valor_extra = valor_extra

    @property
    def nombre(self):
      return self._nombre
    
    @property
    def valor(self):
      return self._valor

    @valor.setter
    def valor(self, valor):
        self._valor = valor
    
    @property
    def valor_extra(self):
        return self._valor_extra
    
    @valor_extra.setter
    def valor_extra(self, valor_extra):
        self.valor_extra = valor_extra

    @nombre.setter
    def nombre(self, nombre):
        self.nombre = nombre
    
    def atacar(self, chipo):
        if(chipo.oponente.tiene_ventaja_sobre(chipo)):
            self.atacar_con_valor(chipo, self.valor + self.valor_extra)
        else:
            self.atacar_con_valor(chipo, self.valor)
    
    @abstractmethod
    def atacar_con_valor(self, chipo, valor):
        pass

    def __str__(self):
        self.nombre
    
    